test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/MatchTest.scala

```diff
@@ -62,4 +62,17 @@ class MatchTest {
     assertEquals("None of those", test(5, 20))
   }
 
+  // #2554
+  @Test def matchWithNonIdentityEndMatch(): Unit = {
+    val foo: Option[Int] = Some(42)
+
+    /* This match generates a value class boxing operation in the matchEnd (in
+     * 2.10 and 2.11).
+     */
+    val result =
+      "foo = " ++ (foo match { case Some(0) => "zero" case _ => "unknown" })
+
+    assertEquals("foo = unknown", result)
+  }
+
```

Comment:
Could you also add your minimized version that does not depend on the Scala lib?